File: handlers/topic_lifecycle.py
import logging
import asyncio
from aiogram import Router, F, Bot
from aiogram.types import Message, CallbackQuery, InlineKeyboardMarkup, InlineKeyboardButton
from database.sqlite_db import SQLiteDatabase
from jules.api import JulesAPIClient
from .topics import send_topic_setup_menu

router = Router()
logger = logging.getLogger(__name__)

setup_states = {}

# Simple in-memory storage for pagination tokens
session_pages = {}

@router.message(F.forum_topic_created)
async def topic_created(message: Message, bot: Bot, db: SQLiteDatabase):
    owner_id = await db.get_setting("owner_id")
    if str(message.from_user.id) != owner_id:
        return

    await db.create_topic(message.message_thread_id)
    setup_states[message.message_thread_id] = {"auto_pr": False}
    await send_topic_setup_menu(bot, message.chat.id, message.message_thread_id, auto_pr=False)



@router.callback_query(F.data.startswith("toggle_pr_"))
async def toggle_pr_cb(callback: CallbackQuery):
    topic_id = callback.message.message_thread_id
    current_val = bool(int(callback.data.split("_")[-1]))
    new_val = not current_val

    state = setup_states.get(topic_id, {"auto_pr": False})
    state["auto_pr"] = new_val
    setup_states[topic_id] = state

    from .topics import get_topic_setup_keyboard
    await callback.message.edit_reply_markup(reply_markup=get_topic_setup_keyboard(auto_pr=state["auto_pr"]))
    await callback.answer()

@router.callback_query(F.data == "cancel_setup")
async def cancel_setup_cb(callback: CallbackQuery, bot: Bot, db: SQLiteDatabase):
    topic_id = callback.message.message_thread_id
    await db.delete_topic(topic_id)
    setup_states.pop(topic_id, None)

    await callback.answer("Deleting topic...")
    try:
        await bot.delete_forum_topic(chat_id=callback.message.chat.id, message_thread_id=topic_id)
    except Exception as e:
        logger.error(f"Failed to delete topic {topic_id}: {e}")
        await callback.message.edit_text("Failed to delete topic. Please delete it manually.")

@router.callback_query(F.data == "proceed_setup_new")
async def proceed_new_cb(callback: CallbackQuery, db: SQLiteDatabase):
    api_key = await db.get_setting("api_key")
    if not api_key:
        await callback.answer("API Key not found. Connect account first.", show_alert=True)
        return

    client = JulesAPIClient(api_key)
    try:
        sources_data = await client.list_sources(page_size=100)
        sources = sources_data.get("sources", [])

        if not sources:
            await callback.answer("No repositories found in your Jules account.", show_alert=True)
            return

        kb = []
        for s in sources:
            short_name = s.get('id', s.get('name', '')).replace('github-', '')
            kb.append([InlineKeyboardButton(text=short_name, callback_data=f"repo_{s['name']}")])

        kb.append([InlineKeyboardButton(text="🔄 Refresh", callback_data="proceed_setup_new")])
        kb.append([InlineKeyboardButton(text="Back", callback_data="back_to_setup")])

        await callback.message.edit_text("Select a repository:", reply_markup=InlineKeyboardMarkup(inline_keyboard=kb))
    except Exception as e:
        logger.error(f"Error fetching sources: {e}")
        await callback.answer("Error fetching repositories.", show_alert=True)

@router.callback_query(F.data.startswith("repo_"))
async def repo_selected_cb(callback: CallbackQuery, db: SQLiteDatabase):
    repo_name = callback.data[5:]
    topic_id = callback.message.message_thread_id

    state = setup_states.get(topic_id, {})
    state["selected_repo"] = repo_name
    setup_states[topic_id] = state

    # Now fetch branches for the repo
    api_key = await db.get_setting("api_key")
    client = JulesAPIClient(api_key)

    try:
        source_data = await client.get_source(repo_name)
        branches = source_data.get("githubRepo", {}).get("branches", [])

        if not branches:
            await callback.answer("No branches found for this repository.", show_alert=True)
            return

        kb = []
        for b in branches:
            branch_name = b.get("displayName")
            kb.append([InlineKeyboardButton(text=branch_name, callback_data=f"branch_{branch_name}")])

        kb.append([InlineKeyboardButton(text="Back", callback_data="proceed_setup_new")])

        await callback.message.edit_text("Select a branch:", reply_markup=InlineKeyboardMarkup(inline_keyboard=kb))
    except Exception as e:
        logger.error(f"Error fetching branches: {e}")
        await callback.answer("Error fetching branches.", show_alert=True)

@router.callback_query(F.data.startswith("branch_"))
async def branch_selected_cb(callback: CallbackQuery, db: SQLiteDatabase):
    branch_name = callback.data[7:]
    topic_id = callback.message.message_thread_id

    state = setup_states.get(topic_id, {})
    state["selected_branch"] = branch_name
    setup_states[topic_id] = state

    repo_name = state.get("selected_repo", "Unknown")
    short_repo = repo_name.split("/")[-1].replace("github-", "")

    await callback.message.edit_text(
        f"Selected Repo: `{short_repo}`\nSelected Branch: `{branch_name}`\n\n"
        "Please send the initialization message (prompt) to start the session.",
        parse_mode="Markdown"
    )
    state["waiting_for_prompt"] = True
    setup_states[topic_id] = state

@router.message(F.text, lambda msg: setup_states.get(msg.message_thread_id, {}).get("waiting_for_prompt", False))
async def prompt_received(message: Message, bot: Bot, db: SQLiteDatabase):
    topic_id = message.message_thread_id
    state = setup_states.get(topic_id, {})

    api_key = await db.get_setting("api_key")
    client = JulesAPIClient(api_key)

    repo_name = state.get("selected_repo")
    branch_name = state.get("selected_branch")
    auto_pr = state.get("auto_pr", False)

    status_msg = await message.answer("Starting session...")

    try:
        session_data = await client.create_session(
            prompt=message.text,
            source=repo_name,
            branch=branch_name,
            auto_pr=auto_pr
        )

        session_id = session_data.get("name")
        await db.update_topic_session(topic_id, session_id)
        await db.update_topic_auto_pr(topic_id, auto_pr)

        short_repo = repo_name.split("/")[-1].replace("github-", "")
        pinned_msg = await bot.send_message(
            chat_id=message.chat.id,
            message_thread_id=topic_id,
            text=f"📌 **Session Active**\nSession ID: `{session_id}`\nRepo: `{short_repo}`\nBranch: `{branch_name}`\nState: `{session_data.get('state', 'QUEUED')}`",
            parse_mode="Markdown"
        )
        await bot.pin_chat_message(message.chat.id, pinned_msg.message_id)
        await db.update_topic_pinned_message(topic_id, pinned_msg.message_id)

        await status_msg.delete()
        setup_states.pop(topic_id, None)

    except Exception as e:
        logger.error(f"Error creating session: {e}")
        await status_msg.edit_text(f"Error creating session: {e}")


@router.callback_query(F.data == "back_to_setup")
async def back_to_setup_cb(callback: CallbackQuery):
    topic_id = callback.message.message_thread_id
    state = setup_states.get(topic_id, {"auto_pr": False})
    
    from .topics import get_topic_setup_keyboard
    await callback.message.edit_text(
        "Please configure this topic:",
        reply_markup=get_topic_setup_keyboard(auto_pr=state.get("auto_pr", False))
    )

File: src/handlers/topic_lifecycle.ts
import { Update } from '@grammyjs/types';
import { router } from '../router';
import { DB } from '../db';
import { callApi } from '../api';
import { JulesAPIClient } from '../jules';
import { sendTopicSetupMenu, getTopicSetupKeyboard } from './topics';

const setupStates: Record<number, any> = {};

router.on((u: Update) => !!u.message?.forum_topic_created, async (update: Update) => {
  const message = update.message!;
  const ownerId = DB.getSetting("owner_id");

  if (String(message.from?.id) !== ownerId) return;

  const topicId = message.message_thread_id;
  if (!topicId) return;

  DB.createTopic(topicId);
  setupStates[topicId] = { auto_pr: false };
  await sendTopicSetupMenu(message.chat.id, topicId, false);
});

router.onCallback('toggle_pr_', async (cb, state) => {
  if (!cb.message) return;
  const topicId = cb.message.message_thread_id;
  if (!topicId) return;

  const currentVal = Boolean(parseInt(cb.data!.split("_").pop() || "0"));
  const newVal = !currentVal;

  const topicState = setupStates[topicId] || { auto_pr: false };
  topicState.auto_pr = newVal;
  setupStates[topicId] = topicState;

  await callApi('editMessageReplyMarkup', {
    chat_id: cb.message.chat.id,
    message_id: cb.message.message_id,
    reply_markup: getTopicSetupKeyboard(topicState.auto_pr)
  });
  await callApi('answerCallbackQuery', { callback_query_id: cb.id });
});

router.onCallback('cancel_setup', async (cb, state) => {
  if (!cb.message || !cb.message.message_thread_id) return;
  const topicId = cb.message.message_thread_id;

  DB.deleteTopic(topicId);
  delete setupStates[topicId];

  await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Deleting topic..." });
  try {
    await callApi('deleteForumTopic', { chat_id: cb.message.chat.id, message_thread_id: topicId });
  } catch (e) {
    console.error(`Failed to delete topic ${topicId}:`, e);
    await callApi('editMessageText', {
      chat_id: cb.message.chat.id,
      message_id: cb.message.message_id,
      text: "Failed to delete topic. Please delete it manually."
    });
  }
});

router.onCallback('proceed_setup_new', async (cb, state) => {
  if (!cb.message) return;
  const apiKey = DB.getSetting("api_key");

  if (!apiKey) {
    await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "API Key not found. Connect account first.", show_alert: true });
    return;
  }

  const client = new JulesAPIClient(apiKey);
  try {
    const sourcesData = await client.listSources(100);
    const sources = sourcesData.sources || [];

    if (sources.length === 0) {
      await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "No repositories found in your Jules account.", show_alert: true });
      return;
    }

    const kb: any[][] = [];
    for (const s of sources) {
      const name = s.id || s.name || '';
      const shortName = name.replace('github-', '');
      kb.push([{ text: shortName, callback_data: `repo_${name}` }]);
    }

    kb.push([{ text: "🔄 Refresh", callback_data: "proceed_setup_new" }]);
    kb.push([{ text: "Back", callback_data: "back_to_setup" }]);

    await callApi('editMessageText', {
      chat_id: cb.message.chat.id,
      message_id: cb.message.message_id,
      text: "Select a repository:",
      reply_markup: { inline_keyboard: kb }
    });
  } catch (e) {
    console.error(`Error fetching sources:`, e);
    await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Error fetching repositories.", show_alert: true });
  }
});

router.onCallback('repo_', async (cb, state) => {
  if (!cb.message || !cb.message.message_thread_id) return;
  const repoName = cb.data!.substring(5);
  const topicId = cb.message.message_thread_id;

  const topicState = setupStates[topicId] || {};
  topicState.selected_repo = repoName;
  setupStates[topicId] = topicState;

  const apiKey = DB.getSetting("api_key");
  if (!apiKey) return;

  const client = new JulesAPIClient(apiKey);
  try {
    const sourceData = await client.getSource(repoName);
    const branches = sourceData.githubRepo?.branches || [];

    if (branches.length === 0) {
      await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "No branches found for this repository.", show_alert: true });
      return;
    }

    const kb: any[][] = [];
    for (const b of branches) {
      const branchName = b.displayName;
      kb.push([{ text: branchName, callback_data: `branch_${branchName}` }]);
    }
    kb.push([{ text: "Back", callback_data: "proceed_setup_new" }]);

    await callApi('editMessageText', {
      chat_id: cb.message.chat.id,
      message_id: cb.message.message_id,
      text: "Select a branch:",
      reply_markup: { inline_keyboard: kb }
    });
  } catch (e) {
    console.error(`Error fetching branches:`, e);
    await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Error fetching branches.", show_alert: true });
  }
});

router.onCallback('branch_', async (cb, state) => {
  if (!cb.message || !cb.message.message_thread_id) return;
  const branchName = cb.data!.substring(7);
  const topicId = cb.message.message_thread_id;

  const topicState = setupStates[topicId] || {};
  topicState.selected_branch = branchName;
  
  const repoName = topicState.selected_repo || "Unknown";
  const shortRepo = repoName.split("/").pop()?.replace("github-", "");

  await callApi('editMessageText', {
    chat_id: cb.message.chat.id,
    message_id: cb.message.message_id,
    text: `Selected Repo: \`${shortRepo}\`\nSelected Branch: \`${branchName}\`\n\nPlease send the initialization message (prompt) to start the session.`,
    parse_mode: "Markdown"
  });

  topicState.waiting_for_prompt = true;
  setupStates[topicId] = topicState;
});

// We handle prompt responses in a generic filter
router.on((u) => {
  if (!u.message || !u.message.text || !u.message.message_thread_id) return false;
  const state = setupStates[u.message.message_thread_id];
  return state && state.waiting_for_prompt === true;
}, async (update) => {
  const message = update.message!;
  const topicId = message.message_thread_id!;
  const topicState = setupStates[topicId] || {};

  const apiKey = DB.getSetting("api_key");
  if (!apiKey) return;
  const client = new JulesAPIClient(apiKey);

  const repoName = topicState.selected_repo;
  const branchName = topicState.selected_branch;
  const autoPr = topicState.auto_pr || false;

  const statusMsg = await callApi<any>('sendMessage', {
    chat_id: message.chat.id,
    message_thread_id: topicId,
    text: "Starting session..."
  });

  try {
    const sessionData = await client.createSession(message.text!, repoName, branchName, autoPr);
    const sessionId = sessionData.name;

    DB.updateTopicSession(topicId, sessionId);
    DB.updateTopicAutoPr(topicId, autoPr);

    const shortRepo = repoName.split("/").pop()?.replace("github-", "") || '';
    const cleanId = sessionId.split("/").pop() || '';
    const julesUrl = sessionData.url || `https://jules.google.com/session/${cleanId}`;

    const autoPrStr = autoPr ? "`ON`" : "`OFF`";
    const stateStr = sessionData.state || 'QUEUED';
    
    const emojiMap: Record<string, string> = {
      "QUEUED": "⏳", "PLANNING": "🧠", "AWAITING_PLAN_APPROVAL": "✋", 
      "AWAITING_USER_FEEDBACK": "💬", "IN_PROGRESS": "🔄", "PAUSED": "⏸️", 
      "FAILED": "❌", "COMPLETED": "✅"
    };
    const stateEmoji = emojiMap[stateStr] || "🔵";

    const richMd = `
# ${sessionId}

| **Status** | **Auto PR** | **Repo** | **Branch** | **State** |
|:---|:---|:---|:---|:---|
| \`Active\` | ${autoPrStr} | \`${shortRepo}\` | \`${branchName}\` | ${stateEmoji} \`${stateStr}\` |
`;
    
    // sendRichMessage is supported in recent Bot APIs. 
    // For raw Telegram API, sendRichMessage isn't broadly standard if not yet rolled out globally or if they don't have access.
    // Assuming they have it or it's similar:
    const pinnedMsg = await callApi<any>('sendRichMessage', {
      chat_id: message.chat.id,
      message_thread_id: topicId,
      rich_message: { markdown: richMd },
      reply_markup: {
        inline_keyboard: [
          [
            { text: "Open Session", url: julesUrl },
            { text: "No PR", callback_data: "no_pr_alert" }
          ]
        ]
      }
    });

    await callApi('pinChatMessage', { chat_id: message.chat.id, message_id: pinnedMsg.message_id });
    DB.updateTopicPinnedMessage(topicId, pinnedMsg.message_id);

    await callApi('deleteMessage', { chat_id: message.chat.id, message_id: statusMsg.message_id });
    delete setupStates[topicId];

  } catch (e) {
    console.error(`Error creating session:`, e);
    await callApi('editMessageText', {
      chat_id: message.chat.id,
      message_id: statusMsg.message_id,
      text: `Error creating session: ${e}`
    });
  }
});

router.onCallback('no_pr_alert', async (cb) => {
  await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "There is no PR yet.", show_alert: true });
});

router.onCallback('back_to_setup', async (cb) => {
  if (!cb.message || !cb.message.message_thread_id) return;
  const topicId = cb.message.message_thread_id;
  const topicState = setupStates[topicId] || { auto_pr: false };

  await callApi('editMessageText', {
    chat_id: cb.message.chat.id,
    message_id: cb.message.message_id,
    text: "Please configure this topic:",
    reply_markup: getTopicSetupKeyboard(topicState.auto_pr)
  });
});
